File: leetcodes/141. Linked List Cycle/linkedListCycle.java
public class Solution {
  public boolean hasCycle(ListNode head) {
    /**
     * two pointer: time O(n) worse case no cycle, spaceO(1) use slow and fast =
     * slow.next.next iterate over the list til null if slow is fast return true
     * else slow.next fast.next.next return false
     * 
     */
    if (head == null || head.next == null)
      return false;
    ListNode slow = head;
    ListNode fast = head.next;
    while (slow != null && fast != null && fast.next != null) {
      if (slow == fast)
        return true;
      slow = slow.next;
      fast = fast.next.next;
    }
    return false;
  }
}

File: leetcodes/141. Linked List Cycle/linkedListCycle.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# time O(n), space O(1) with two pointers

class Solution:
    def hasCycle(self, head: Optional[ListNode]) -> bool:
      if not head or not head.next or not head.val:
        return False
      fast = head.next
      slow = head
      while not slow and not fast and not fast.next:
        if fast == slow:
          return True
        slow = slow.next
        fast = fast.next.next
      return False

# compare node == node don't use val
#[1,1,1,1]
# -1
#